pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -1823,3 +1879,109 @@ func makeListenerKey(l gatev1alpha2.Listener) string {
 }
+
+func updateHTTPRouteStatus(ctx context.Context, client Client, gateway *gatev1alpha2.Gateway, listener gatev1alpha2.Listener, route *gatev1alpha2.HTTPRoute) error {
+	status := route.Status.RouteStatus.DeepCopy()
+	spec := route.Spec.CommonRouteSpec.DeepCopy()
+	return client.UpdateHTTPRouteStatus(route, makeRouteStatus(ctx, *status, *spec, route.ObjectMeta, gateway, listener))
+}
+
+func updateTCPRouteStatus(ctx context.Context, client Client, gateway *gatev1alpha2.Gateway, listener gatev1alpha2.Listener, route *gatev1alpha2.TCPRoute) error {
+	status := route.Status.RouteStatus.DeepCopy()
+	spec := route.Spec.CommonRouteSpec.DeepCopy()
+	return client.UpdateTCPRouteStatus(route, makeRouteStatus(ctx, *status, *spec, route.ObjectMeta, gateway, listener))
+}
+
+func updateTLSRouteStatus(ctx context.Context, client Client, gateway *gatev1alpha2.Gateway, listener gatev1alpha2.Listener, route *gatev1alpha2.TLSRoute) error {
+	status := route.Status.RouteStatus.DeepCopy()
+	spec := route.Spec.CommonRouteSpec.DeepCopy()
+	return client.UpdateTLSRouteStatus(route, makeRouteStatus(ctx, *status, *spec, route.ObjectMeta, gateway, listener))
+}
+
+func makeRouteStatus(ctx context.Context, routeStatus gatev1alpha2.RouteStatus, routeSpec gatev1alpha2.CommonRouteSpec, routeMeta metav1.ObjectMeta, gateway *gatev1alpha2.Gateway, listener gatev1alpha2.Listener) gatev1alpha2.RouteStatus {
+	var routeParentStatus *gatev1alpha2.RouteParentStatus
+
+	// Check if we need to update an existing parent reference
+	for i, parent := range routeStatus.Parents {
+		if parent.ControllerName != controllerName {
+			continue
+		}
+
+		parentRef := parent.ParentRef
+
+		if parentRef.Group == nil || *parentRef.Group != gatev1alpha2.GroupName {
+			continue
+		}
+		if parentRef.Kind == nil || *parentRef.Kind != kindGateway {
+			continue
+		}
+		if parentRef.SectionName != nil && *parentRef.SectionName != listener.Name {
+			continue
+		}
+
+		namespace := routeMeta.Namespace
+		if parentRef.Namespace != nil {
+			namespace = string(*parentRef.Namespace)
+		}
+
+		if namespace == gateway.Namespace && string(parentRef.Name) == gateway.Name {
+			log.Ctx(ctx).Debug().Msgf("Found existing parent reference for route %s/%s", routeMeta.Namespace, routeMeta.Name)
+			routeParentStatus = &routeStatus.Parents[i]
+		}
+	}
+
+	// No existing parent reference was found - create a new one
+	if routeParentStatus == nil {
+		log.Ctx(ctx).Debug().Msgf("Adding parent reference for route %s/%s", routeMeta.Namespace, routeMeta.Name)
+
+		group := gatev1alpha2.Group(gatev1alpha2.GroupName)
+		kind := gatev1alpha2.Kind(kindGateway)
+		namespace := gatev1alpha2.Namespace(gateway.Namespace)
+
+		// Only set the section name if the route requests it
+		var sectionName *gatev1alpha2.SectionName
+		for _, p := range routeSpec.ParentRefs {
+			if p.SectionName != nil && *p.SectionName == listener.Name {
+				sectionName = p.SectionName
+			}
+		}
```

Comment:
> I think the current approach would not allow specifying an empty string to yield that the Route has been attached to all listeners of a Gateway.

If i understood the spec correctly then I think it's still possible to specify an empty string. The motivation here was to set the section name explicitly on the HTTPRouteStatus object if the HTTProute spec requests a particular gateway listener object.

If the HTTPRoute does not care about which listener it gets attached to, the section name is left empty in the manifest file which means that the route can be attached to any listener and so the route parent status object would have an empty string there too if that makes sense?

Here's an example of what I'm thinking:
```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: Gateway
metadata:
  name: public-gw
  namespace: gateway-ingress
spec:
  gatewayClassName: traefik
  listeners:
    - name: http
      protocol: HTTP
      port: 80
      ...
    - name: https # <-- this is the section name
      protocol: HTTPS
      port: 443
      ...
---
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: prometheus
  namespace: monitoring
spec:
  parentRefs:
    - name: public-gw
      namespace: gateway-ingress
      sectionName: https # <-- only attach the route to HTTPs listener. if left empty it'll get
                         # attached to both - port 80 and port 443 and the corresponding
                         # status object will have an empty section name too
  rules: ...
```

> the case of a Gateway having 3 Listeners, if the route is attached only to two listeners, the empty string value doesn’t seem appropriate but a single sectionName is not good as well.

I think that's how the spec wants it. If you want a particular listener you must specify that in the HTTPRoute object (the sectioname). Otherwise it'll just get attached to whatever was matched.

Reply:
Sure, I think we are on the same line regarding the attachment.
My question was more about the `RouteStatus` update.
As of today, the place where it is built doesn't allow taking into account the multiple listener attachment, and hence, specifying an empty string.
If I'm not overlooking this, to do that, it is needed to build the `RouteStatus` after every listener has been processed.

Also, it is still unclear to me if multiple [RouteParentStatus](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.RouteParentStatus) can be added for the same Gateway in a RouteStatus.

Reply:
I've read the spec more times than I'd like to admit, but I think the current implementation is accurate. The status object is from the perspective of the route. If the route spec has an empty `sectionName` it means it can be attached to *any* listener when deciding whether to allow attaching the route to the gateway. I don't think we need to update the status object with the listener that accepted it. It should be set to whatever route requested because that is the object it is referencing.

The same goes for when the route spec has set a `sectionName` and/or `port`. Both values must match in order for the route to be considered successfully attached to the gateway in which case the status object will also contain the `sectionName` and/or `port` that the *route* requested.